Release: RST 4.6

ecbland opened the pull request:
# Release RST 4.6 

- [ ] PATCH
- [X] MINOR
- [ ] MAJOR

## Deadline: 19 July 2021
Please comment in the Pull Request conversation if you need additional time to test the release.

## Testing

- [ ] Mac OSx
- [x] Linux deb - Debian, Ubuntu, Mint
- [x] Linux rpm - OpenSuse, Fedora, CentOS

*Do any binaries/libraries require special attention during testing?*

Below are the major updates. Of course these have already been tested, but feel free to try to "break" them :smile: 

- `make_fit -fitacf-version 3.0` - fixed "divide by zero" bug that affects noise level determination for BOREALIS radars (#420 . Added Shepherd (2017) elevation angle algorithm
- `make_grid` - see #373 
- `fit_speck_removal` - new binary, see #400 
- `datdump` - new binary, see #376
- `fov_plot` - new functionality, see #440 
- IDL DLM for MLT_v2 - see #437 
- [Documentation](https://radar-software-toolkit-rst.readthedocs.io/en/develop/index.html) - lots of updates, comments welcome. 

### How to test

Install git, and then run:

```
git clone https://github.com/SuperDARN/rst
cd rst
git checkout release/rst-4.6
make.build && make.code
```

Alternatively, download a [zip archive of the release branch](https://github.com/SuperDARN/rst/archive/refs/heads/release/rst-4.6.zip) and then run `make.build && make.code` as usual. You may need to update some environment variables for this to work. 

From there, please use RST as normal and play with the new features listed above. 

## Checklist

- [X] Update `radar.dat` -- not required this time
- [x] Update hardware files (if required)
- [X] Update `.rst.version`
- [x] Update author list in `.zenodo.json`. **@ksterne and I will provide a draft author list ASAP**
- [x] Write a list of major updates for the release description
 
## Extra Notes

All hardware files have been updated to match the new file format on the `hdw` repo. 
The DCE and DCN radars have updated information (see https://github.com/SuperDARN/hdw/pull/1). 


## Approving the release

After testing, please state which operating system you used and what you tested.

If you found no bugs during testing, please *approve* the PR:

1. Click on `files changed` top right
2. Click on `start review` green top right  button
3. In the comment box, state which operating system you used for testing, and any other helpful information
4. Select `approve` and `submit review`

ksterne commented:
Excellent writeup on this pull request @ecbland!  In the next week or so, I'll give things a closer look, but tested this for compilation on Ubuntu 20 and passes this test.

pasha-ponomarenko reviewed (approved):
I briefly tested the requested list of changed/added routines on OpenSuse Leap 15.3 (the most recent version):
- make_fit -fitacf-version 3.0 -- tested both BOREALIS low-level noise fix and switching to Simon's script for elevation.
- make_grid -- tested error messages
- fit_speck_removal -- made sure that it removes the "salt-and-pepper" noise
- datdump -- tested it with and without -d key
- fov_plot -- tested the comma-separated option, plotting both highlighted and "empty" FOVs plus the hemisphere determination feature
- IDL DLM for MLT_v2  -- tested the new keyword

ksterne commented:
The draft authorship list is up with 8941985.  Not sure if folks got the notice on the commit added, so leaving a comment here.

ecbland commented on docs/dev/copyright_license.md:
```diff
@@ -0,0 +1,206 @@
+When developing in RST you are granting permission for your code to be licensed under the GPL. This will be ok in almost all situations. Exceptions may arise if your employer wants to make your program into proprietary software, or if your funding agency has restrictions on the publication of research outputs (e.g. defense contracts). If you suspect that you won't be allowed to contribute code to ...
+
+!!! IMPORTANT
+    Please make sure to review the [license]((https://www.gnu.org/licenses/gpl-3.0.en.html)), and check with your employer/funding agency that you have permission to distribute your code under the GPLv3. 
```
This fixes a broken link

```suggestion
    Please make sure to review the [license](https://www.gnu.org/licenses/gpl-3.0.en.html), and check with your employer/funding agency that you have permission to distribute your code under the GPLv3. 
```

ecbland commented on docs/dev/copyright_license.md:
```diff
@@ -0,0 +1,206 @@
+
+RST was originally developed at the Johns Hopkins University/Applied Physics Laboratory (JHU/APL). Around 2012, JHU/APL granted permission for RST to be re-licensed so that the SuperDARN community could continue maintaining the software collaboratively. This process has caused confusion over the years, since even scientists who are expert programmers may not be familiar with the intricacies of so ...
+
+A major source of confusion around the RST license is whether it was intended to be licensed under the [GNU General Public License](https://www.gnu.org/licenses/gpl-3.0.en.html) (GPL), or the [GNU __Lesser__ General Public License](https://www.gnu.org/licenses/lgpl-3.0.en.html) (LGPL). This confusion has arisen because, at some point, the LGPL license notice was attached to most of the RST source ...
```
Just fixed the last sentence to say "was" instead of "will be"

```suggestion
A major source of confusion around the RST license is whether it was intended to be licensed under the [GNU General Public License](https://www.gnu.org/licenses/gpl-3.0.en.html) (GPL), or the [GNU __Lesser__ General Public License](https://www.gnu.org/licenses/lgpl-3.0.en.html) (LGPL). This confusion has arisen because, at some point, the LGPL license notice was attached to most of the RST source code. The `AstAlg` library is the exception to this, which has been [clearly marked](https://github.com/SuperDARN/rst-archive/blob/rst.3.1/codebase/analysis/src.lib/astalg/astalg.1.2/LICENSE.txt) with a GPL disclaimer since it was first added to the RST in v3.1. Since `AstAlg` is clearly licensed under the GPL, it follows that the whole of the RST should also be licensed under the GPL. It is possible that the LGPL disclaimer text was added to the remaining RST source code in error since it is very similar to the GPL disclaimer text. To add to this confusion, RST was not distributed with any license file for several releases (v3.5 to v4.3 inclusive). The GPL license file was added in [v4.3.1](https://doi.org/10.5281/zenodo.3634732). The license information across the whole package was corrected in RST4.6.
```

ksterne replied:
Woohoo! Feels good to put things in the past.

ecbland reviewed (commented):
Two tiny changes in the copyright/license documentation

egthomas commented on codebase/superdarn/src.idl/app/examples.1.2/readcnvmap.pro:
```diff
@@ -2,30 +2,24 @@
-; 
-; 
 ;
-; A Simply demonstration program for reading cnvmap files
```
It appears the description for each of these IDL example routines was inadvertently deleted when changing the license/copyright header information.

ksterne replied:
Thanks for the catch here @egthomas!  Looks like that might be a whoops back on 7280cf2.  Is there something special with a Simply demo program or is that a typo and might read better as: "A simple demonstration program"?  Not sure if this is something specific to IDL or not.

egthomas replied:
@ksterne agreed that was likely a typo, and your suggested change ("A simple demonstration program for ...") sounds appropriate.

ecbland replied:
Fixed this with https://github.com/SuperDARN/rst/pull/441/commits/a694599068849626e6112ba99550f62d27078fc5

egthomas commented on codebase/superdarn/src.idl/app/makefit.1.5/calcphires.pro:
```diff
@@ -1,31 +1,26 @@
+; 
+; RST is free software: you can redistribute it and/or modify
+; it under the terms of the GNU General Public License as published by
 ; ---------------
```
It does not seem like the license/copyright header information was correctly imported into the example IDL makefit app files.

ksterne replied:
Looks like the last 2 lines here should be removed.  I should be able to address these in the next few hours.

ksterne replied:
These files have been resolved with ce449d1.

egthomas commented on codebase/superdarn/src.idl/lib/efield.1.4/efield.pro:
```diff
@@ -2,30 +2,26 @@
+; Modifications:
 ; 
-;
-;
-; PublicFunctions
+; RST is free software: you can redistribute it and/or modify
+; it under the terms of the GNU General Public License as published by
 ; ---------------
 ;
 ; EFieldRead        
```
Similar to above, the "Public Functions" list header appears to have been inadvertently removed from this and several other IDL library files.

ksterne replied:
Should be able to take care of this one as well.  Thanks again for catching these!

ksterne replied:
This is resolved with ce449d1.

ksterne replied:
Whoops, this specific one is actually fixed with f4842ac.

egthomas commented on codebase/superdarn/src.idl/lib/main.1.25/snd.pro:
```diff
@@ -2,7 +2,24 @@
+; 
+; Modifications:
+; 
+; Functions
```
```suggestion
```

This extra "Functions" label should not have been inserted here.

ksterne replied:
Is this the only file in the directory that needs this change?

egthomas commented on codebase/superdarn/src.idl/lib/main.1.25/rprm.pro:
```diff
@@ -2,30 +2,24 @@
-; Public Functions
-; ----------------
+; Functions
 ;
```
Similar to above, the "Public Functions" list header and underline format has been inadvertently lost from the original file.

ksterne replied:
Nevermind, I think this answer the question.  It's not the same change to be done, but does touch all of the files.

ksterne replied:
This should be resolved by 2382044.

gighen74 reviewed (approved):
I tested RST v 4.6 in the following OS: Ubuntu 16.04.5 LTS

1) Installation performed following the instructions in here: https://github.com/SuperDARN/rst/pull/441. 
SUGGESTION: In the installation instructions it would be better to specify that the enviroment variables and the .bashrc shell settings should be set before doing make.build && make.code, otherwise the compiling could fail...

2) I have tested the release performing a full processing from rawacfs to cnvmap on a test case, which was already processed by dr. Federica Marcucci using version 4.4.1. The test case is a 2-hours data chunk during a more extended event (geomagnetic storm of 25-26 August 2018), in detail: 2018-08-25 18:00-20:00 UT, Northern Hemisphere.
List of input files used:
-rwxrwx--- 1 root vboxsf 76422720 lug  7 11:07 20180825.1801.00.ade.rawacf
-rwxrwx--- 1 root vboxsf 88078500 lug  7 11:09 20180825.1801.00.bks.rawacf
-rwxrwx--- 1 root vboxsf 58137600 lug  7 11:09 20180825.1801.00.cly.rawacf
-rwxrwx--- 1 root vboxsf 74392707 lug  7 11:10 20180825.1801.00.cve.rawacf
-rwxrwx--- 1 root vboxsf 74431034 lug  7 11:10 20180825.1801.00.cvw.rawacf
-rwxrwx--- 1 root vboxsf 30432000 lug  7 11:10 20180825.1801.00.gbr.rawacf
-rwxrwx--- 1 root vboxsf 82252800 lug  7 11:11 20180825.1801.00.han.rawacf
-rwxrwx--- 1 root vboxsf 82206720 lug  7 11:12 20180825.1801.00.hkw.rawacf
-rwxrwx--- 1 root vboxsf 80747520 lug  7 11:13 20180825.1801.00.hok.rawacf
-rwxrwx--- 1 root vboxsf 58137600 lug  7 11:12 20180825.1801.00.inv.rawacf
-rwxrwx--- 1 root vboxsf 55574400 lug  7 11:14 20180825.1801.00.kap.rawacf
-rwxrwx--- 1 root vboxsf 55637760 lug  7 11:14 20180825.1801.00.ksr.rawacf
-rwxrwx--- 1 root vboxsf 82337280 lug  7 11:16 20180825.1801.00.lyr.rawacf
-rwxrwx--- 1 root vboxsf 44004480 lug  7 11:15 20180825.1801.00.pgr.rawacf
-rwxrwx--- 1 root vboxsf 82268160 lug  7 11:16 20180825.1801.00.pyk.rawacf
-rwxrwx--- 1 root vboxsf 44004480 lug  7 11:16 20180825.1801.00.rkn.rawacf
-rwxrwx--- 1 root vboxsf 44004480 lug  7 11:17 20180825.1801.00.sas.rawacf
-rwxrwx--- 1 root vboxsf 34447284 lug  7 11:21 20180825.1923.31.kod.rawacf
-rwxrwx--- 1 root vboxsf 11498100 lug  7 11:21 20180825.1945.25.cve.rawacf
-rwxrwx--- 1 root vboxsf 11498100 lug  7 11:21 20180825.1945.25.cvw.rawacf
NOTE: for the "stereo" radars (ade, ksr, kod), only channel "a" ("d" for kod) has been used and  the files above have been renamed.

List of commands run:

>> Creating fitacfs using fitacf v 2.5:
for i in *.rawacf; do for j in $(basename $i .rawacf); do make_fit -vb -fitacf-version 2.5 $i > ../dest_path/$j.fitacf; done; done
>> Creating grdmaps extended forcing a fixed scan lenght of 120 seconds: 
for i in *.fitacf; do for j in $(basename $i .fitacf); do make_grid -vb -xtd -tl 120 $i > grd/$j.grdmap; done; done 
>> Combine grids:
combine_grid -vb 20180825.1801.00.ade.grdmap 20180825.1801.00.bks.grdmap 20180825.1801.00.cly.grdmap 20180825.1801.00.cve.grdmap 20180825.1801.00.cvw.grdmap 20180825.1801.00.gbr.grdmap 20180825.1801.00.han.grdmap 20180825.1801.00.hkw.grdmap 20180825.1801.00.hok.grdmap 20180825.1801.00.inv.grdmap 20180825.1801.00.kap.grdmap 20180825.1801.00.ksr.grdmap 20180825.1801.00.lyr.grdmap 20180825.1801.00.pgr.grdmap 20180825.1801.00.pyk.grdmap 20180825.1801.00.rkn.grdmap 20180825.1801.00.sas.grdmap 20180825.1923.31.kod.grdmap 20180825.1945.25.cve.grdmap 20180825.1945.25.cvw.grdmap > 20180825_1800_2000_N.grdmap

>> Produce cnvmap though the usual steps; the IMF+VEL+Kp TXT file is attached:
[OMNI_IMF_kp_18-20.txt](https://github.com/SuperDARN/rst/files/6814943/OMNI_IMF_kp_18-20.txt)

map_grd -vb ../grd/20180825_1800_2000_N.grdmap > 20180825_1800_2000_N.empty.cnvmap
  
 map_addhmb -vb -nodef 20180825_1800_2000_N.empty.cnvmap > 20180825_1800_2000_N.hmb.cnvmap
  
 map_addimf -vb -d 0 -if OMNI_IMF_kp_18-20.txt 20180825_1800_2000_N.hmb.cnvmap > 20180825_1800_2000_N.imf.cnvmap

map_addmodel -vb -ts18_kp -o 8 -d low 20180825_1800_2000_N.imf.cnvmap > 20180825_1800_2000_N.model.cnvmap

map_fit -vb 20180825_1800_2000_N.model.cnvmap > 20180825_1800_2000_N.cnvmap

>> I then converted the cnvmap in an old version ASCII map file, as I'm used to visual inspect them and I have some IDL routine ready for extracting the series of the most interesting parameters. I have in fact extracted the series of the cross polar cap potential and compared with one obtained by Federica Marcucci with v. 4.4.1 and the agreement is very good.

3) Verified the Phi0 is correctly computed for dce and dcn with fitacf v 3.0

Unfortunately I'm leaving for vacations, so that I cannot test the Release further. As far as I could verify, It seems to work very well.

ecbland commented:
Thanks to everyone who has been testing this release, and helping to identify and correct some last-minute bugs/errors. 

**Has anyone been able to test installation on MacOS?**

----------

I've just added @gighen74 to the author list (https://github.com/SuperDARN/rst/pull/441/commits/64c473324964a164c1671aaa94d18b459e814c5f) -- thanks for your help and for the detailed description of the testing you did :100: 

I've also been testing this branch on Linux Mint 20 for a few weeks without any problems. I'm mostly using `make_fit` and various plotting routines.

ecbland commented:
Here's a brief description of this release that can be used on Zenodo and elsewhere. Please me know if anything is missing :smile: 

-----------------

New features in version 4.6 of the Radar Software Toolkit (RST) include:

- Routine for removing non-gaussian noise/interference from `fitacf` files (`fit_speck_removal`)
- Routine to display the contents of old-format `dat` files (`datdump`)
- Shepherd (2017) elevation angle algorithm added to FITACF3.0
- Ability to plot multiple fields of view with `fov_plot`
- Added missing `mlt2mlon` keyword to MLT_v2 IDL/DLM code
- `make_grid` detects and concatenates multiple input files automatically (deprecates `-c` flag)
- Check that the search noise is nonzero before using it to replace the skynoise in FITACF3.0
- Fixed bugs in plotting libraries, `cdf` file reading, `make_grid` and `trim_raw`
- Improved compliance with GPLv3 license requirements
- Documentation updates

The RST is actively developed and maintained by the SuperDARN Data Analysis Working Group (https://superdarn.github.io/dawg/).

egthomas commented on codebase/superdarn/src.idl/app/examples.1.2/readcfit.pro:
```diff
@@ -1,3 +1,30 @@
+;readcfit.pro
+;==============
+;Author: R.J.Barnes
+
+;Copyright (c) 2012 The Johns Hopkins University/Applied Physics Laboratory
```
Maybe this goes against a style guide I am unaware of, but I personally find comments in IDL (and other languages) to be more legible when there is a space between the comment character and text.

ecbland replied:
We don't have a style guide, but I that's fair enough about the spacing.

mts299 commented:
Looks good got it installed on OpenSuse Leap 15.2 and tested make_fit and datdump

ecbland commented:
I think this is basically ready to go, but I just got a request from the DCN PI to update the hardware file with a new TDIFF value. New version is on its way

ksterne commented:
At thought at some point I or someone had added Ubuntu 20 dependencies to the install instructions, but I don't see them there.  I've just run through an Ubuntu 20 install and could add them here. Thoughts?  Maybe I'm remembering adding the Ubuntu 18 dependencies...

ecbland commented:
> @ksterne At thought at some point I or someone had added Ubuntu 20 dependencies to the install instructions, but I don't see them there. I've just run through an Ubuntu 20 install and could add them here. Thoughts? Maybe I'm remembering adding the Ubuntu 18 dependencies...

Are they any different to Ubuntu 18?

ksterne commented:
Very very slightly.  Ubuntu 20 doesn't ship with make or gcc it seems (or at least my install didn't), so you have to add those in.  The install line is slightly different too: `apt install` replaces `apt-get install`.

ecbland commented:
> Very very slightly. Ubuntu 20 doesn't ship with make or gcc it seems (or at least my install didn't), so you have to add those in. The install line is slightly different too: `apt install` replaces `apt-get install`.

Ok, go for it then! We're still waiting on the DCN hardware file anyway. Is it just `build-essential` that's needed? That should include both `gcc` and `make`

ksterne commented:
OK, I've got the Ubuntu 20 install notes with 398e22c.  Hopefully the formatting looks OK on the read-the-docs site, don't think we've got a release branch version of read-the-docs going, do we?

ecbland commented:
> Hopefully the formatting looks OK on the read-the-docs site, don't think we've got a release branch version of read-the-docs going, do we?

I tried to get this working a few weeks ago, but readthedocs couldn't see the release branch. I just checked again, and magically it's working...here's the link: https://radar-software-toolkit-rst.readthedocs.io/en/release-rst-4.6/

I think the table of contents on the homepage needs some work...I'll take a look at that tomorrow.

ksterne commented:
Ok, thanks for pointing me to that...I didn't try very hard to figure out how to convert the `/` to a URL character.  Looks like the table needed some help to create some empty cells, so I've fixed that with 7957a91.  Otherwise, the `build-essential` was under 18.04.

Agreed, the table of contents on the home page fairly expanded when I pull over content from the RFCs.  I at least tried to keep it organized.  Maybe some of the sections can be condensed with expand/collapse measure features much like the left-hand navigation bar.  If it's too much, I wouldn't worry about it, we can refine that for a future release.

ecbland commented:
@ksterne Can we make the release now?

ksterne commented:
Yes, though I'm not in a great position to hit the button to merge until Thursday (too many meetings).  Feel free to move it along without me.  I've just reviewed [the checklist](https://github.com/SuperDARN/dawg/blob/doc/release-guidelines/docs/current_work/RST_release_guidelines.md#release-checklist) and added a draft release to the repo with the summary of changes that was provided.  So I think we're through step 9 (we've probably indirectly reminded people on testing).  For step 11, we'll want to edit the draft release to point to `master` and then hit publish release.  I realize "draft" in this step is confusing as this was written before you could create 'draft releases'.  I'll edit that here shortly.
